secure_software_development_fundamentals.md

```diff
@@ -3949,7 +3949,7 @@ A particularly nasty security problem in computer systems is *insecure random nu
+Insecure random number generation is an especially serious problem in Internet of Things (IoT) device software. One 2021 report found that about 35 billion IoT devices had disastrous security vulnerabilities due to insecure cryptographic random number generation. This is in part because many IoT software developers directly call hardware random number generators (they shouldn’t do that), but even worse, they ignored error return codes from those generators (and they definitely shouldn’t do that). Hardware random number generators typically have easily-exceeded generation rates, so errors are often very likely. Just paying attention to the error code from the hardware isn’t really enough, though.  Hardware sometimes fails, and if the software just hangs in that case, the IoT device may be too unreliable to compete. Sample code for accessing the random number generator hardware is often insecure (it shows how to get data, but not how to use it correctly). Correctly using the hardware directly can be quite difficult, for example, the LPC 54628’s user manual page number 1,106 (of 1,152) notes, in a convoluted way, that after reading a random number from its hardware you must read and throw away the next 32 values.
 
 Software developers for IoT devices should not access the hardware registers directly, but should instead call well-crafted CSPRG generators that correctly use hardware sources (preferably multiple sources) as inputs into their internal entropy pool. In most cases IoT developers should use an IoT operating system that includes a CSPRG implementation that is correctly seeded from multiple hardware sources, and simply check to see if it appears to be carefully written for security. Where that’s not practical, use a well-crafted and analyzed CSPRNG library that includes correct software to extract random values from your hardware; do not implement your own crypto unless you’re an expert in cryptography. IoT software developers should also run statistical tests on their random number generation mechanism to ensure that they’re random, because this is an especially common problem in IoT devices. For more details, see [You're Doing IoT RNG](https://labs.bishopfox.com/tech-blog/youre-doing-iot-rng) ([presentation](https://www.youtube.com/watch?v=Zuqw0-jZh9Y)) by Dan Petro and Allan Cecil, a 2021 DEF CON presentation.
```

Comment:
Would be good to explain what CSPRG is. 

Software developers for IoT devices should not access the hardware registers directly, but should instead call well-crafted CSPRG (Cryptographically secure pseudorandom number generator, https://en.wikipedia.org/wiki/Cryptographically-secure_pseudorandom_number_generator) generators that correctly use hardware sources (preferably multiple sources) as inputs into their internal entropy pool.

Reply:
Okay. Here's my attempt to define CSPRNG without diving too much into the weeds: "A good CSPRNG prevents practically predicting the next output given past outputs and it also prevents revealing past outputs if its internal state is compromised." I'll also link to the CSPRNG Wikipedia article.

Reply:
Looks good to me, thanks!